### PR TITLE
fix: exporting async module as library `type:  "module"`

### LINKED ISSUE #13995: Async module library with `type: "module"` exports `undefined`
# Bug report

**What is the current behavior?**

```
Welcome to Node.js v16.6.1.
Type ".help" for more information.
> await import(`./dist/bundle.js`)
[Module: null prototype] { a: undefined }
```

**If the current behavior is a bug, please provide the steps to reproduce.**

package.json
```json
{"type":"module"}
```

```js
// entry.js

export const a = await Promise.resolve(1);
```

```js
// webpack.config.js

export default {
	mode: "development",
	entry: "./entry.js",
	devtool: false,
	output: {
		path: new URL("./dist", import.meta.url).pathname,
		filename: "bundle.js",
		library: {
			type: "module"
		}
	},
	experiments: {
		topLevelAwait: true,
		outputModule: true
	}
};
```

**What is the expected behavior?**

```
Welcome to Node.js v16.6.1.
Type ".help" for more information.
> await import(`./dist/bundle.js`)
[Module: null prototype] { a: 1 }
```

**Other relevant information:**
webpack version:  5.50.0
Node.js version: 16.6.1
Operating System:  macos big sur
Additional tools:

### PR DESCRIPTION
**Pull request by yujunjung**
**What kind of change does this PR introduce?**

fixes #13995

**Did you add tests for your changes?**
Yes

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**

Output module is using `top-level await`.

### CODE REVIEW

**Review by alexander-akait — CHANGES_REQUESTED** [review]
Please add test

File: lib/library/ModuleLibraryPlugin.js [hunk 1]
```diff
@@ -78,13 +78,20 @@ class ModuleLibraryPlugin extends AbstractLibraryPlugin {
 		const result = new ConcatSource(source);
 		const exportsInfo = moduleGraph.getExportsInfo(module);
 		const exports = [];
+		const isAsync = moduleGraph.isAsync(module);
+		if (isAsync)
+			result.add(
+				`var __webpack_async_exports__ = await __webpack_exports__;\n`
+			);
```
Comment (sokra) [inline lib/library/ModuleLibraryPlugin.js:85]:
```suggestion
		if (isAsync) {
			result.add(
				`__webpack_exports__ = await __webpack_exports__;\n`
			);
		}
```

File: lib/library/ModuleLibraryPlugin.js [hunk 2]
```diff
@@ -78,13 +78,20 @@ class ModuleLibraryPlugin extends AbstractLibraryPlugin {
...
 		for (const exportInfo of exportsInfo.orderedExports) {
 			if (!exportInfo.provided) continue;
 			const varName = `__webpack_exports__${Template.toIdentifier(
 				exportInfo.name
 			)}`;
 			result.add(
-				`var ${varName} = __webpack_exports__${propertyAccess([
+				`var ${varName} = ${
+					isAsync ? "__webpack_async_exports__" : "__webpack_exports__"
+				}${propertyAccess([
```
Comment (sokra) [inline lib/library/ModuleLibraryPlugin.js:94]:
```suggestion
				`var ${varName} = __webpack_exports__${propertyAccess([
```

**Review by sokra — CHANGES_REQUESTED** [review]

File: test/ProgressPlugin.test.js [hunk 3]
```diff
@@ -79,6 +79,7 @@ const RunCompilerAsync = compiler =>
 	});
 
 describe("ProgressPlugin", function () {
+	jest.setTimeout(10000);
```
Comment (alexander-akait) [inline test/ProgressPlugin.test.js:82]: Should work without this...

Comment (yujunjung) [reply test/ProgressPlugin.test.js]: reverted

Comment (yujunjung) [reply test/ProgressPlugin.test.js]: failed randomly again in ProgressPlugin
this seems ci error...

**Review by alexander-akait — APPROVED** [review]

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**